Firmware/src/modules/tools/temperaturecontrol/max31855.h

```diff
@@ -0,0 +1,76 @@
+/*
+      this file is part of smoothie (http://smoothieware.org/). the motion control part is heavily based on grbl (https://github.com/simen/grbl).
+      smoothie is free software: you can redistribute it and/or modify it under the terms of the gnu general public license as published by the free software foundation, either version 3 of the license, or (at your option) any later version.
+      smoothie is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose. see the gnu general public license for more details.
+      you should have received a copy of the gnu general public license along with smoothie. if not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#ifndef max31855_h
+#define max31855_h
+
+#include "TempSensor.h"
+
+#include "Pin.h"
+#include "RingBuffer.h"
+
+class Max31855 : public TempSensor
+{
+    public:
+
+        /**
+         * @brief   Initialize target sensor
+         * @param   Nothing
+         * @return  Nothing
+         * @note    The module is initialized with a null pointer
+         */
+        Max31855();
+
+        /**
+         * @brief   Deinitialize target sensor
+         * @param   Nothing
+         * @return  Nothing
+         */
+        ~Max31855() {};
+
+        /**
+         * @brief   Configure the module using the parameters from the config file
+         * @param   cr          : target config file object
+         * @param   m           : target section map of the config file
+         * @return  false       : the module failed to be configured
+         * @return  true        : the module is configured and loaded successfully
+         */
+        bool configure(ConfigReader& cr, ConfigReader::section_map_t& m);
+
+        /**
+         * @brief   Output average temperature value of the sensor
+         * @param   Nothing
+         * @return  Average of the last acquired temperature values
+         */
+        float get_temperature();
+
+        /**
+         * @brief   Acquire temperature data from the sensor
+         * @param   os          : output stream
+         * @return  Nothing
+         * @note    Raw data is acquired and converted to temperature in Nuttx
+         */
+        void get_raw(OutputStream& os);
+
+        //The following variables need to be public as Nuttx functions need access to them
```

Comment:
Not acceptable design.

Reply:
This is a thread implementation i could use. My purpose was to create the max31855 sub module without changing much of the main modules such as temperature control. However, I had to launch thread on Temperaturecontrol module and i used common methods present on tempsensor.h for that thread. Since there is no on_idle function anymore which was previously used on smoothiev1, i had to use the avaliable get_raw method with that outputstream argument, this is why i think i could use it for print messages on uart. If you're not ok with this, what would you suggest for implementing this thread instead?

Reply:
get_raw is not used to read temps, it was only used for the debug M305 command.
I am not sure I understand what you are doing here. In V1 we would have used on_idle, however as you noted there is no on_idle in V2, and command_ctx callback is not suitable either as it only gets called every 200ms or slower. You would need to start a thread to read the SPI, and have the thread sleep most of the time. I do not really like that as it uses a lot of memory to do that. (This is what I did in the Player.cpp module, but I am not totally happy with it). When the thread reads the SPI it puts it in the buffer, then the main code can read that buffer as needed.